### PR TITLE
Cleaned up AmqpValue, MessageProperties, MessageHeader and Message to be more C++-like.

### PR DESCRIPTION
**Pull request by LarryOsterman**
# Updated AMQP Value types to make them feel closer to C++ types and be more useful.

Also scrubbed the entire Value surface to ensure it complies with the AMQP specification.

## Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

**Review by gearama — APPROVED** [review]

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 1]
```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
 
-    // Headers.
-    bool IsHeaderTypeByDescriptor() const;
-    Header GetHeaderFromValue() const;
+  protected:
+    UniqueAmqpValueHandle m_value;
+  };
 
-    static AmqpValue CreateHeader(Header const& header);
+  namespace _detail {
```
Comment (LarryOsterman) [inline sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:371]: I'm not 100% sure if this should be in the `_detail` namespace or not. The public methods of this class are intended to be interactable from customers, because they are implementing the relevant STL functions that allow these types to be interacted with as if they were the corresponding STL type. But if they're in the `_detail` namespace, then they won't appear in the ApiView.

Comment (antkmsft) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:371]: We have the similar situation in `DateTime` - it is ok for customers to write `DateTime::duration`, but not ok to write `Azure::_detail::Clock::duration`. I think this is the similar case.
Another similar case is probably public typedeffing `_internal::UniqueHandle<PublicType>` - and it is ok for customers to use it, but not to write `_internal::...`

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 2]
```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
+
+    template <typename T, typename ThisType> class AmqpCollectionBase {
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:373]: 👍

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 3]
```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
+
+    bool operator==(AmqpSymbol const& that) const { return m_value == that.m_value; }
+    bool operator==(const decltype(m_value)::value_type* const that) const
+    {
+      return m_value == that;
+    }
   };
 
-  /** @brief An AmqpMap represents an AMQP "map" type. which maps a key to a value.
+  class AmqpTimestamp {
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:563]: BTW, I think it is good to just review all class and struct declarations in core-amqp for having `final` and depending on that having/not having virtual/protected destructor.
```suggestion
  class AmqpTimestamp final {
```

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp [hunk 4]
```diff
@@ -188,6 +245,7 @@ TEST_F(TestValues, TestChar)
   {
     AmqpValue value{AmqpValue::CreateChar(37)};
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:246]: Should we also write a few test cases with different unicode char literals, at least for illustrative puropses?
https://stackoverflow.com/questions/6796157/unicode-encoding-for-string-literals-in-c11

For examples, "\u2764\uFE0F" is the same thing as "\xE2\x9D\xA4\xEF\xB8\x8F", they mean the same character (heart emoji), and it probably can fit into char32_t.

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp [hunk 5]
```diff
@@ -200,69 +258,102 @@ TEST_F(TestValues, TestTimestamp)
   {
     std::chrono::milliseconds timeNow{std::chrono::duration_cast<std::chrono::milliseconds>(
         std::chrono::system_clock::now().time_since_epoch())};
-    AmqpValue value{AmqpValue::CreateTimestamp(timeNow)};
-    EXPECT_EQ(timeNow, value.GetTimestamp());
+    AmqpTimestamp value{timeNow};
+    EXPECT_EQ(static_cast<std::chrono::milliseconds>(value), timeNow);
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:262]: BTW, should it be `operator std::chrono::milliseconds`, or `operator `std::chrono::duration`?

File: sdk/core/azure-core-amqp/src/models/amqp_value.cpp [hunk 6]
```diff
@@ -671,9 +738,32 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
+  }
+  std::ostream& operator<<(std::ostream& os, AmqpSymbol const& value)
+  {
+    // Let the AmqpValue specialization handle serialization of the array.
+    AmqpValue arrayValue(value);
+    os << arrayValue;
+    return os;
+  }
+
   bool AmqpValue::IsNull() const
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/src/models/amqp_value.cpp:764]: should there also be ==/!= operators for `(AmqpValue const& val, std::nullptr_t)` + `(std::nullptr_t, AmqpValue const& val)` calling `val.IsNull()`?

That may allow semantics like `if (amqpValue == nullptr)`, if we want that.

File: sdk/core/azure-core-amqp/src/models/message_source.cpp [hunk 7]
```diff
@@ -24,159 +30,154 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models { namespace
...
 
-  MessageSource::~MessageSource()
-  {
-    if (m_source != nullptr)
-    {
-      source_destroy(m_source);
-      m_source = nullptr;
-    }
-  }
+  MessageSource::~MessageSource() {}
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/src/models/message_source.cpp:68]:
```suggestion
  MessageSource::~MessageSource() = default;
```

File: sdk/core/azure-core-amqp/src/models/message_target.cpp [hunk 8]
```diff
@@ -24,160 +30,153 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models { namespace
...
 
-  MessageTarget::~MessageTarget()
-  {
-    if (m_target != nullptr)
-    {
-      target_destroy(m_target);
-      m_target = nullptr;
-    }
-  }
+  MessageTarget::~MessageTarget() {}
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/src/models/message_target.cpp:66]:
```suggestion
  MessageTarget::~MessageTarget() = default;
```

**Review by antkmsft — APPROVED** [review]

File: sdk/core/azure-core-amqp/test/ut/amqp_header_tests.cpp [hunk 9]
```diff
@@ -16,62 +16,81 @@ class TestHeaders : public testing::Test {
 TEST_F(TestHeaders, SimpleCreate)
 {
   {
-    Header header;
+    MessageHeader header;
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/test/ut/amqp_header_tests.cpp:19]: The line is there to verify that you can statically construct a MessageHeader with no parameters.

Most/all of my unit tests that are designed to test a type start with something like 
```cpp
{
   Foo foo;
}
```

From the simplest case (declare a value of type Foo), then you move to building other constructs on top of that.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 10]
```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
-    // Char
-    static AmqpValue CreateChar(uint32_t value);
-    uint32_t GetChar() const;
+    AmqpTimestamp AsTimestamp() const;
 
-    // Timestamps
-    static AmqpValue CreateTimestamp(std::chrono::milliseconds value);
-    std::chrono::milliseconds GetTimestamp() const;
+    // Get Char value.
+    std::uint32_t GetChar() const;
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:355]: MSVC doesn't seem to define `std::char32_t` :(. That's quite annoying, to be honest - I'd love to use it here.

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp [hunk 11]
```diff
@@ -38,148 +39,204 @@ TEST_F(TestValues, SimpleCreate)
...
     EXPECT_EQ(AmqpValueType::String, value.GetType());
     EXPECT_EQ(std::string("Fred"), fredP);
+    EXPECT_TRUE(AmqpValue() < value);
+  }
+
+  {
+    Azure::Core::Uuid uuid = Azure::Core::Uuid::CreateUuid();
+    AmqpValue value(uuid);
+    EXPECT_EQ(AmqpValueType::Uuid, value.GetType());
+    EXPECT_EQ(uuid.ToString(), static_cast<Azure::Core::Uuid>(value).ToString());
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:130]: Possibly, but beyond the charter of this pull request or feature. Personally, I'd wait until we have a customer request for it, and this feature technically does not need it.

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp [hunk 12]
```diff
@@ -38,148 +39,204 @@ TEST_F(TestValues, SimpleCreate)
...
     AmqpValue boolValue{false};
-    EXPECT_ANY_THROW(boolValue.GetListItemCount());
+    EXPECT_ANY_THROW(AmqpList list(boolValue));
   }
   // Put some things in the list.
   {
-    AmqpValue list1{AmqpValue::CreateList()};
-    list1.SetListItemCount(4);
-    EXPECT_EQ(4, list1.GetListItemCount());
+    const AmqpList list1{123, 23.97f, "ABCD", static_cast<char>('a')};
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:161]: That fix isn't sufficient, because there is no 'char' ctor for AmqpValue.

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp [hunk 13]
```diff
@@ -200,69 +258,102 @@ TEST_F(TestValues, TestTimestamp)
...
+    EXPECT_ANY_THROW(boolValue.AsSymbol());
   }
 }
 
 TEST_F(TestValues, TestCompositeValue)
 {
   {
-    AmqpValue value{AmqpValue::CreateComposite("My Composite Type", 5)};
+    AmqpComposite value;
+    //    AmqpComposite value("My Composite Type", {1, 2, 5.5, "ABC", 5});
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:292]: Sort-of. It didn't compile yesterday, but the changes I made today seem to work.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_header.hpp [hunk 14]
```diff
@@ -3,49 +3,75 @@
...
 
-    bool operator==(Header const& that) const
-    {
-      return (this->IsDurable() == that.IsDurable())
-          && (this->GetDeliveryCount() == that.GetDeliveryCount())
-          && (this->GetTimeToLive() == that.GetTimeToLive())
-          && (this->IsFirstAcquirer() == that.IsFirstAcquirer());
-    }
-    friend std::ostream& operator<<(std::ostream&, Header const&);
+    friend std::ostream& operator<<(std::ostream&, MessageHeader const&);
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_header.hpp:72]: It needs to be declared, just not with a friend operator.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_message.hpp [hunk 15]
```diff
@@ -18,59 +36,187 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
+     * - A single AMQP Value.
+     *
+     * This method sets the body of the message to a single AMQP value. See [Amqp
+     * Value](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-amqp-value)
+     * for more information.
+     *
+     * @param bodyValue - a single value binary data.
+     *
+     */
+    void SetBody(AmqpValue const& bodyValue);
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_message.hpp:191]: That would be ... problematic, I believe. The thing about `AmqpList`, `AmqpBinaryData` and friends is that they explicitly are *not* `AmqpValue`s. 

An `AmqpValue` is a container for `AMQP_VALUE` type. But `AmqpList` and friends are types which bring C++ semantics to the AMQP non-value types. They function as C++ containers and *NOT* `AMQP_VALUE` types.

My biggest concern here is that there are some fairly significant semantic differences between the C++ container types and the AMQP_VALUE types. By converting the AMQP_VALUE types directly to the corresponding C++ container types, the types behave like C++ types, but if I use the uAMQP AMQP_VALUE functions, the semantics would get muddled.

I may reconsider this at some point, but personally, I like the paradigm of "It's a native type until it cannot be a native type" (and the corollary  "it gets turned into a native type as soon as possible").

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_message.hpp [hunk 16]
```diff
@@ -18,59 +36,187 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
-    void AddBodyAmqpData(BinaryData binaryData);
-
-    BinaryData GetBodyAmqpData(size_t index) const;
-    size_t GetBodyAmqpDataCount() const;
-
-    void SetBodyAmqpValue(AmqpValue value);
-    AmqpValue GetBodyAmqpValue() const;
+    operator UniqueMessageHandle() const;
 
     friend std::ostream& operator<<(std::ostream&, Message const&);
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_message.hpp:215]: All the friend declarations are now just free method declarations.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_properties.hpp [hunk 17]
```diff
@@ -4,68 +4,52 @@
...
+    using type
+        = Azure::Core::_internal::BasicUniqueHandle<PROPERTIES_INSTANCE_TAG, FreeAmqpProperties>;
+  };
+}}} // namespace Azure::Core::_internal
+
+namespace Azure { namespace Core { namespace Amqp { namespace Models {
 
-  private:
-    PROPERTIES_INSTANCE_TAG* m_properties{};
+  using UniquePropertiesHandle = Azure::Core::_internal::UniqueHandle<PROPERTIES_INSTANCE_TAG>;
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_properties.hpp:23]: I'm confused. `UniquePropertiesHandle` *is* a typedef.

If you have suggestions about how I could implement C# style "assembly local" types in C++, I'd love to hear them. That's what the `operator UniqueXxxHandle` and `Xxx(XXX_INSTANCE_TAG)` functions logically are.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_properties.hpp [hunk 18]
```diff
@@ -4,68 +4,52 @@
...
 
+  class MessageProperties final {
     // uAMQP interop functions.
   public:
-    Properties(PROPERTIES_INSTANCE_TAG* properties) : m_properties(properties) {}
-    operator PROPERTIES_INSTANCE_TAG*() const { return m_properties; }
+    MessageProperties(PROPERTIES_INSTANCE_TAG* properties); //: m_properties(properties) {}
+    operator UniquePropertiesHandle() const;
 
   public:
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_properties.hpp:31]: The second "public:" would better be "internal", but that isn't a C++ language feature.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 19]
```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
+
+    /** @brief Construct an AMQP ubyte value, an 8 bit unsigned integer.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.3](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-ubyte).
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(std::uint8_t value);
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:119]: Already done.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 20]
```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
...
+    operator std::int32_t() const;
+    operator std::uint64_t() const;
+    operator std::int64_t() const;
     operator float() const;
-    operator float();
     operator double() const;
-    operator double();
-    operator BinaryData() const;
-    operator BinaryData();
     explicit operator std::string() const;
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:338]: I made it explicit because I found several bugs because of implicit conversions to std::string. To remove those bugs, I made the std::string conversions explicit.

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp (hunk 4 above)
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:246]: I believe that "\u2764\uFE0F" is not a UTF-32 encoded value, I believe it's 2 UTF-16 codepoints. The UTF-32 encoded value for the heart emoji is [0x1f48c](https://unicode.org/emoji/charts/full-emoji-list.html#1f48c)., so it is the equivalent of calling `CreateChar(0x1f48c)`. I'm not sure I see a semantic difference between `CreateChar(37)` and `CreateChar(0x1f48c)`.

The underlying AMQP "Char" type is defined as a 32 bit unsigned integer value. There are no additional semantics defined for it (and most importantly, there are no conversion semantics defined for AMQP Char values, so it doesn't make sense to add tests for it).

File: sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp (hunk 5 above)
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/test/ut/amqp_value_tests.cpp:262]: The AMQP "timestamp" type is defined as a value in milliseconds.

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_properties.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_properties.hpp]: If we really need it, we can use the thing that is available to us - i.e. `friend`/internal classes etc.
